server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSet.scala

```diff
@@ -76,25 +98,89 @@ case class EmailCreationRequest(mailboxIds: MailboxIds,
+          .map(multipartBuilder => {
+            builder.setBody(multipartBuilder)
+            builder.build
+          })
+      } else {
+        Right(builder.setBody(maybeHtmlBody.getOrElse(""), SubType.HTML_SUBTYPE, StandardCharsets.UTF_8).build)
+      }
+    })
+  }
+
+  private def createMultipartWithAttachments(maybeHtmlBody: Option[String],
+                                             attachments: List[Attachment],
+                                             attachmentManager: AttachmentManager,
+                                             attachmentContentLoader: AttachmentContentLoader,
+                                             mailboxSession: MailboxSession): Either[ErrorCreatingAttachmentException, MultipartBuilder] = {
+    val multipartBuilder = MultipartBuilder.create(SubType.MIXED_SUBTYPE)
+    val bodypartBuilder = BodyPartBuilder.create()
+    val inlineAttachments = attachments.filter(_.isInline)
+    val otherAttachments = attachments.filter(!_.isInline)
+
+    if (inlineAttachments.nonEmpty) {
+      val relatedMultipartBuilder: MultipartBuilder = MultipartBuilder.create(SubType.RELATED_SUBTYPE)
+      relatedMultipartBuilder.addBodyPart(bodypartBuilder.setBody(maybeHtmlBody.getOrElse(""), SubType.HTML_SUBTYPE, StandardCharsets.UTF_8))
+
+      inlineAttachments
+        .flatMap(attachment => getAttachmentMetadata(attachment, attachmentManager, mailboxSession))
+        .foreach(attachmentMetadata => relatedMultipartBuilder.addBodyPart(
+          bodypartBuilder.setBody(attachmentContentLoader.load(attachmentMetadata, mailboxSession).readAllBytes(), attachmentMetadata.getType.asString())
+            .setContentDisposition("inline")))
+
+      multipartBuilder
+        .addBodyPart(BodyPartBuilder.create.setBody(maybeHtmlBody.getOrElse(""), SubType.PLAIN_SUBTYPE, StandardCharsets.UTF_8))
+        .build
+    }
+
+    val maybeAttachments: Either[ErrorCreatingAttachmentException, List[(AttachmentMetadata, Array[Byte])]] = otherAttachments
+      .flatMap(attachment => getAttachmentMetadata(attachment, attachmentManager, mailboxSession))
+      .map(attachmentMetadata => loadAttachment(attachmentMetadata, attachmentContentLoader, mailboxSession))
+      .sequence
+
+    maybeAttachments.map(list => {
+      list.foldLeft(multipartBuilder) {
+        case (builder, (attachment, content)) =>
+          builder.addBodyPart(bodypartBuilder.setBody(content, attachment.getType.asString())
+            .setContentDisposition("attachment"))
+          builder
+      }
     })
+  }
+
+  private def getAttachmentMetadata(attachment: Attachment, attachmentManager: AttachmentManager, mailboxSession: MailboxSession): Option[AttachmentMetadata] = {
+    Try(attachmentManager.getAttachment(AttachmentId.from(attachment.blobId.value.toString), mailboxSession))
+      .fold(_ => None, attachment => Some(attachment))
```

Comment:
toOption does this.

BTW using an either will prevent to "loose" the exception...